tests/abi/abi.c

```diff
@@ -0,0 +1,175 @@
+    CALL_A_FUNCTION(FUNCTION_NAME)                                           \
+    /* Restore the stack */                                                  \
+    RESTORE_STACK(STACK_VALUE_COUNT)                                         \
+    /* Restore all the registers */                                          \
+    REVERSE_FOR_EACH(POP_REGISTER, REGISTER_LIST)                            \
+    :: "X"(Generated)                                                        \
+    : COMMA_SEPARATED_FOR_EACH(STRINGIFY, REGISTER_LIST), "memory"           \
+  );                                                                         \
+                                                                             \
+  free(Generated);                                                           \
+}
+/* clang-format on */
+
+#ifdef __APPLE__
+#define FUNCTION_NAME_PREFIX _
+#else
+#define FUNCTION_NAME_PREFIX
+#endif
+
+#ifdef REG_
+#undef REG_
+#endif
+#define REG_ "%"
+
+#ifdef LOC_
+#undef LOC_
+#endif
+#define LOC_ "%" STRINGIFY(FRAME_POINTER)
+
+/* clang-format off */
+#define NAKED_FUNCTIONS __asm__ (                                          \
+  /* Define the first function */                                          \
+  STRINGIFY(FUNCTION_NAME_PREFIX) "getTheValue:\n"                         \
+  UPDATE_FRAME_POINTER                                                     \
+  ALLOCATE_ON_HEAP(FRAME_POINTER, (GENERATED_COUNT) * (REGISTER_SIZE))     \
+  CALL_A_FUNCTION_WITH_FRAME_POINTER(generate)                             \
+  CALL_A_FUNCTION_WITH_FRAME_POINTER(dump)                                 \
+                                                                           \
+  FOR_EACH(PUSH_REGISTER, REGISTER_LIST)                                   \
+  REPEAT(PUSH_CONSTANT, STACK_VALUE_COUNT, REGISTER_COUNT, FIRST_REGISTER) \
+  INDEXED_FOR_EACH(FILL_REGISTER, REGISTER_LIST)                           \
+                                                                           \
+  FIX_RETURN_ADDRESS((GENERATED_COUNT + 1) * (REGISTER_SIZE))              \
+  RETURN_FROM_FUNCTION                                                     \
+                                                                           \
+  /* Define the second function */                                         \
+  STRINGIFY(FUNCTION_NAME_PREFIX) "restoreTheState:\n"                     \
+  POP_REGISTER(rbp)                                                        \
+  RESTORE_STACK(STACK_VALUE_COUNT)                                         \
+  REVERSE_FOR_EACH(POP_REGISTER, REGISTER_LIST)                            \
+  RESTORE_STACK(2)                                                         \
+  /* FREE_ON_HEAP(FRAME_POINTER) // A known memory leak. TODO: fix. */     \
+  PUSH_REGISTER(rbp)                                                       \
+  RETURN_FROM_FUNCTION                                                     \
+)
+NAKED_FUNCTIONS;
```

Comment:
* [ ] This piece of code is quite obscure. Document what's going on. What are `getTheValue` and `restoreTheState`?
* [ ] Why `#define X` and use it right away? Can't you just drop the macro and use its content right away?

Reply:
1. Yes, right, I'm sorry. That's my bad. When I managed to get this ~~mess~~ piece of code working, I felt really tired so I decided to do it the next morning. And then forgot and moved on to some other stuff. I'll look into documenting this.
2. That's a trick to avoid our convention checker complaining :) It really doesn't like the first line ending in `(`.

Reply:
P. S. For where `getTheValue` and `restoreTheState` functions are used you can take a look [here](https://github.com/revng/revng-qa/pull/4/files#diff-69068270dabcdde6ac2fe63aa22ce2d4972973a1b5eee6fae36d18f2cc27c1abR224) and [here](https://github.com/revng/revng-qa/pull/4/files#diff-69068270dabcdde6ac2fe63aa22ce2d4972973a1b5eee6fae36d18f2cc27c1abR231)

P. P. S. I know that technically calling a casted function pointer is UB, but considering that the values of every register and a bunch of stack is literally filled in with predetermined random stuff - that's the least of our problems there :D

Reply:
1. What are they doing though? Seems like a sort of `setjmp`/`longjmp`, you want to use those, maybe?
2. Add a comment or whatever, let's not add one single more macro then necessary `:)`

Reply:
They are called from a c-function. `getTheValue` one sets the state up (fills registers and stack). The [test function](https://github.com/revng/revng-qa/pull/4/files#diff-69068270dabcdde6ac2fe63aa22ce2d4972973a1b5eee6fae36d18f2cc27c1abR225) calls it and pretends to receive some value - which is then `printf`'ed. After which it calls `restoreTheState` which makes sure all the register values and the stack are restored to the way that would have been if none of this happened.

Reply:
You could:
1. `malloc` an area
2. `setjmp`
3. Pivot the stack to the `malloc`'d area
4. Call the function
5. Print the return value
6. `longjmp`

This way you won't need `restoreTheState` nor the part of `getValue` that saves it!

Reply:
Wait, right, how have this never came up when I was trying to find a cleaner way to do it. This actually makes a whole lot of sense. The only question is how portable is `setjmp/longjmp` pair? Since it seems to be part of linux kernel.

Reply:
Done.